src/assets/StylusAsset.js

```diff
@@ -50,18 +50,56 @@ class StylusAsset extends Asset {
   }
 }
 
-async function createEvaluator(asset) {
-  const Evaluator = await localRequire(
-    'stylus/lib/visitor/evaluator',
-    asset.name
+async function getDependencies(code, asset) {
+  const [Parser, DepsResolver] = await Promise.all(
+    ['parser', 'visitor/deps-resolver'].map(dep =>
+      localRequire('stylus/lib/' + dep, asset.name)
+    )
   );
-  const utils = await localRequire('stylus/lib/utils', asset.name);
-  const resolver = new Resolver(
+
+  let deps = new Map();
+  let resolver = new Resolver(
     Object.assign({}, asset.options, {
       extensions: ['.styl', '.css']
     })
   );
 
+  class ImportVisitor extends DepsResolver {
+    visitImport(imported) {
+      let path = imported.path.first.string;
+
+      if (!deps.has(path)) {
+        deps.set(path, resolver.resolve(path, asset.name).then(m => m.path));
+      }
+    }
+  }
+
+  let options = {
+    cache: true,
+    filename: asset.name
+  };
+  let parser = new Parser(code, options);
```

Comment:
unfortunately I think this will result in the code being parsed twice for each file: once here, and once when actually processing the code later. Do you think there is a way they could share the same AST?

Reply:
I tested by putting some logs in Stylus' code and (ironically) `cache: true` broke the Parser cache (the cache is disabled if it isn't `'memory'`).

I fixed it and now it'll parse in `getDependencies` and use the cached AST in `generate`.